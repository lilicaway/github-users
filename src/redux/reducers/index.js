import { combineReducers } from 'redux';
import usersReducer, * as gitHubUsersApi from './gitHubUsers';
import currentUserReducer, * as currentUserApi from './gitHubCurrentUserDetails';

const reducers = combineReducers({
  users: usersReducer,
  currentUser: currentUserReducer,
});

export default reducers;

// Selectors

export const getUsers = state => {
  return gitHubUsersApi.getUsers(state.users);
};

export const getUsersNextLink = state => {
  return gitHubUsersApi.getUsersNextLink(state.users);
};

export const getUsersLoadingState = state => {
  return gitHubUsersApi.getLoadingState(state.users);
};

export const getUsersLoadingErrorMessage = state => {
  return gitHubUsersApi.getErrorMessage(state.users);
};

export const getCurrentUser = state => {
  return currentUserApi.getCurrentUser(state.currentUser);
};

export const getCurrentUserLoadingState = state => {
  return currentUserApi.getCurrentUserLoadingState(state.currentUser);
};

export const getCurrentUserLoadingErrorMessage = state => {
  return currentUserApi.getCurrentUserLoadingErrorMessage(state.currentUser);
};
