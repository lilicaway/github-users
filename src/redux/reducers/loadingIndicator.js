import * as actionType from '../actions';

export const LoadingState = Object.freeze({
  INITIAL: 'INITIAL',
  LOADING: 'LOADING',
  COMPLETED: 'COMPLETED',
  ERROR: 'ERROR',
});

const createLoadingIndicatorReducer = actionPrefix => {
  return (
    state = { loadingState: LoadingState.INITIAL, errorMessage: '' },
    action
  ) => {
    switch (action.type) {
      case `${actionPrefix}_${actionType.DATA_LOADER_LOADING}`:
        return {
          ...state,
          loadingState: LoadingState.LOADING,
          errorMessage: '',
        };
      case `${actionPrefix}_${actionType.DATA_LOADER_COMPLETED}`:
        return {
          ...state,
          loadingState: LoadingState.COMPLETED,
          errorMessage: '',
        };
      case `${actionPrefix}_${actionType.DATA_LOADER_ERROR}`:
        return {
          ...state,
          loadingState: LoadingState.ERROR,
          errorMessage: action.payload,
        };
      default:
        return state;
    }
  };
};

export default createLoadingIndicatorReducer;

// Selectors

export const getLoadingState = state => {
  return state.loadingState;
};

export const getLoadingDataErrorMessage = state => {
  return state.errorMessage;
};
