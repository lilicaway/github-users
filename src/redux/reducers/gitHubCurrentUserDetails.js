import * as actionType from '../actions';
import { combineReducers } from 'redux';
import createLoadingIndicatorReducer, * as loadingIndicatorApi from './loadingIndicator';

const userReducer = (state = {}, action) => {
  switch (action.type) {
    case actionType.SET_CURRENT_USER:
      return action.payload;
    default:
      return state;
  }
};

const currentUserReducer = combineReducers({
  user: userReducer,
  loadingIndicator: createLoadingIndicatorReducer('CURRENT_USER')
});

export default currentUserReducer;

// Selectors

export class CurrentUserApi {
  constructor(state) {
    this.state = state;
  }

  getCurrentUser() {
    return this.state.user;
  }

  getLoadingState() {
    return loadingIndicatorApi.getLoadingState(this.state.loadingIndicator);
  }

  getErrorMessage() {
    return loadingIndicatorApi.getLoadingDataErrorMessage(
      this.state.loadingIndicator
    );
  }
}
