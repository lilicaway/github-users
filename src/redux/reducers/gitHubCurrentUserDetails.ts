import { combineReducers, Reducer } from 'redux';
import { GitHubUser } from '../../types/GitHubUser';
import {
  ActionType,
  DataLoaderActionSubtype,
  GitHubUsersAction
} from '../actions';
import createLoadingIndicatorReducer, * as loadingIndicatorApi from './loadingIndicator';

export type UserState = GitHubUser | null;

const userReducer: Reducer<UserState, GitHubUsersAction> = (
  state = null,
  action
) => {
  switch (action.type) {
    case ActionType.SET_CURRENT_USER:
      return action.payload;
    default:
      return state;
  }
};

interface CurrentUserState {
  user: UserState;
  loadingIndicator: loadingIndicatorApi.LoadingIndicatorState;
}

const currentUserReducer: Reducer<
  CurrentUserState,
  GitHubUsersAction
> = combineReducers({
  user: userReducer,
  loadingIndicator: createLoadingIndicatorReducer(
    DataLoaderActionSubtype.CURRENT_USER
  )
});

export default currentUserReducer;

// Selectors

export class CurrentUserApi {
  constructor(private readonly state: CurrentUserState) {}

  public getCurrentUser(): GitHubUser | undefined {
    return this.state.user || undefined;
  }

  public getLoadingState(): loadingIndicatorApi.LoadingState {
    return loadingIndicatorApi.getLoadingState(this.state.loadingIndicator);
  }

  public getErrorMessage(): string {
    return loadingIndicatorApi.getLoadingDataErrorMessage(
      this.state.loadingIndicator
    );
  }
}
