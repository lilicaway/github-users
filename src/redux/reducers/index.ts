import { combineReducers, Reducer } from 'redux';
import { GitHubUsersAction } from '../actions';
import currentUserReducer, {
  CurrentUserApi,
  CurrentUserState
} from './gitHubCurrentUserDetails';
import usersReducer, { GitHubUsersApi, UsersState } from './gitHubUsers';

export interface AppState {
  users: UsersState;
  currentUser: CurrentUserState;
}

const reducers: Reducer<AppState, GitHubUsersAction> = combineReducers({
  users: usersReducer,
  currentUser: currentUserReducer
});

export default reducers;

// Selectors

export const gitHubUsersApi = (state: AppState) => {
  return new GitHubUsersApi(state.users);
};

export const currentUserApi = (state: AppState) => {
  return new CurrentUserApi(state.currentUser);
};
