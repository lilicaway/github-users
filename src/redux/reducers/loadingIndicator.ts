import { Reducer } from 'redux';
import {
  ActionType,
  DataLoaderActionSubtype,
  GitHubUsersAction
} from '../actions';

export enum LoadingState {
  INITIAL = 'INITIAL',
  LOADING = 'LOADING',
  COMPLETED = 'COMPLETED',
  ERROR = 'ERROR'
}

export interface LoadingIndicatorState {
  loadingState: LoadingState;
  errorMessage: string;
}

const createLoadingIndicatorReducer = (
  actionSubtype: DataLoaderActionSubtype
): Reducer<LoadingIndicatorState, GitHubUsersAction> => {
  return (
    state = {
      loadingState: LoadingState.INITIAL,
      errorMessage: ''
    },
    action
  ) => {
    switch (action.type) {
      case ActionType.DATA_LOADER_LOADING:
        if (action.subType === actionSubtype) {
          return {
            ...state,
            loadingState: LoadingState.LOADING,
            errorMessage: ''
          };
        }
        break;
      case ActionType.DATA_LOADER_COMPLETED:
        if (action.subType === actionSubtype) {
          return {
            ...state,
            loadingState: LoadingState.COMPLETED,
            errorMessage: ''
          };
        }
        break;
      case ActionType.DATA_LOADER_ERROR:
        if (action.subType === actionSubtype) {
          return {
            ...state,
            loadingState: LoadingState.ERROR,
            errorMessage: action.payload
          };
        }
        break;
    }
    return state;
  };
};

export default createLoadingIndicatorReducer;

// Selectors

export const getLoadingState = (state: LoadingIndicatorState): LoadingState => {
  return state.loadingState;
};

export const getLoadingDataErrorMessage = (
  state: LoadingIndicatorState
): string => {
  return state.errorMessage;
};
