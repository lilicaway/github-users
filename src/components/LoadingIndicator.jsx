import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Alert } from 'react-bootstrap';
import { LoadingState } from '../redux/reducers/loadingIndicator';

class LoadingIndicator extends Component {
  static propTypes = {
    loadingState: PropTypes.string.isRequired,
    errorMessage: PropTypes.string.isRequired
  };

  render() {
    const { loadingState, errorMessage } = this.props;
    switch (loadingState) {
      case LoadingState.LOADING:
        return (
          <i
            className="fa fa-spinner fa-spin"
            style={{ fontSize: 24 + 'px' }}
          />
        );
      case LoadingState.ERROR:
        return (
          <Alert bsStyle="danger">
            <strong>Error!</strong> {errorMessage}.
          </Alert>
        );
      default:
        return <div />;
    }
  }
}

export default LoadingIndicator;
export const LoadingIndicatorForTest = LoadingIndicator;
