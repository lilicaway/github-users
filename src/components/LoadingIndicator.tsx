import * as React from 'react';
import { Alert } from 'react-bootstrap';
import { LoadingState } from '../redux/reducers/loadingIndicator';

interface Props {
  loadingState: string;
  errorMessage: string;
}

const LoadingIndicator: React.SFC<Props> = props => {
  const { loadingState, errorMessage } = props;
  switch (loadingState) {
    case LoadingState.LOADING:
      return (
        <i className="fa fa-spinner fa-spin" style={{ fontSize: 24 + 'px' }} />
      );
    case LoadingState.ERROR:
      return (
        <Alert bsStyle="danger">
          <strong>Error!</strong> {errorMessage}.
        </Alert>
      );
    default:
      return <div />;
  }
};

export default LoadingIndicator;
export const LoadingIndicatorForTest = LoadingIndicator;
