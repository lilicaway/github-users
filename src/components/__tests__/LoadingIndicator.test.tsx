import * as Enzyme from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import * as React from 'react';
import { Alert } from 'react-bootstrap';
import { LoadingState } from '../../redux/reducers/loadingIndicator';
import LoadingIndicator from '../LoadingIndicator';

Enzyme.configure({ adapter: new Adapter() });

const expectSpinnerToShow = (
  component: Enzyme.ShallowWrapper,
  showing: boolean
) => {
  expect(component.find('.fa.fa-spinner.fa-spin').exists()).toBe(showing);
};

test('LoadingIndicator renders initial state', () => {
  const component = Enzyme.shallow(
    <LoadingIndicator loadingState={LoadingState.INITIAL} errorMessage="" />
  );
  expectSpinnerToShow(component, false);
  expect(component).toMatchSnapshot();
});

test('LoadingIndicator renders loading state', () => {
  const component = Enzyme.shallow(
    <LoadingIndicator loadingState={LoadingState.LOADING} errorMessage="" />
  );
  expect(component).toMatchSnapshot();
  expectSpinnerToShow(component, true);
});

test('LoadingIndicator renders error state', () => {
  const expectedErrorMessage = 'Some Error Message';
  const component = Enzyme.shallow(
    <LoadingIndicator
      loadingState={LoadingState.ERROR}
      errorMessage={expectedErrorMessage}
    />
  );
  expect(component).toMatchSnapshot();
  expectSpinnerToShow(component, false);
  expect(component.find(Alert).exists()).toBe(true);
});

test('LoadingIndicator renders completed state', () => {
  const component = Enzyme.shallow(
    <LoadingIndicator loadingState={LoadingState.COMPLETED} errorMessage="" />
  );
  expectSpinnerToShow(component, false);
  expect(component).toMatchSnapshot();
});
