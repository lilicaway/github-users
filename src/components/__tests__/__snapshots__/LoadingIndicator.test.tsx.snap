// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadingIndicator renders completed state 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoadingIndicator
    errorMessage=""
    loadingState="COMPLETED"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`LoadingIndicator renders error state 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoadingIndicator
    errorMessage="Some Error Message"
    loadingState="ERROR"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "bsClass": "alert",
      "bsStyle": "danger",
      "children": Array [
        <strong>
          Error!
</strong>,
        " ",
        "Some Error Message",
        ".",
      ],
      "closeLabel": "Close alert",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Error!",
        },
        "ref": null,
        "rendered": "Error!",
        "type": "strong",
      },
      " ",
      "Some Error Message",
      ".",
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bsClass": "alert",
        "bsStyle": "danger",
        "children": Array [
          <strong>
            Error!
</strong>,
          " ",
          "Some Error Message",
          ".",
        ],
        "closeLabel": "Close alert",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Error!",
          },
          "ref": null,
          "rendered": "Error!",
          "type": "strong",
        },
        " ",
        "Some Error Message",
        ".",
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`LoadingIndicator renders initial state 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoadingIndicator
    errorMessage=""
    loadingState="INITIAL"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {},
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`LoadingIndicator renders loading state 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoadingIndicator
    errorMessage=""
    loadingState="LOADING"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "className": "fa fa-spinner fa-spin",
      "style": Object {
        "fontSize": "24px",
      },
    },
    "ref": null,
    "rendered": null,
    "type": "i",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "className": "fa fa-spinner fa-spin",
        "style": Object {
          "fontSize": "24px",
        },
      },
      "ref": null,
      "rendered": null,
      "type": "i",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
